### PR TITLE
[Transform] Check for zero-param operators in LiftTransformParams

### PR DESCRIPTION
**Pull request by Lunderberg**
Prior to this commit, `LiftTransformParams` would extract out all variable binding that have no runtime dependencies.  As a result, expressions such as `R.zeros([16], "int32")` would be extracted out into the parameter transformation, even though they do not depend on any parameters.

This commit updates `LiftTransformParams` to only output variables that depend on at least one compile-time parameter.

The unit test for this functionality also found that `relax::Call` was erroneously calling `MarkGraphNode` in `SEqualReduce` and `SHashReduce`.  This should only be called for nodes that have have reference equality, such as `relax::Var`, and not for composite objects.  This caused erroneous failures in the unit test when two instances of `R.zeros([16], "int32")` were being compared by reference equality in `StructuralEqual`.  These extra calls to `MarkGraphNode` have been removed.

### CODE REVIEW

**Comment by Lunderberg** [comment]
This PR depends on changes made in https://github.com/apache/tvm/pull/16594, and is marked as a draft until it lands.

**Comment by vinx13** [comment]
`As a result, expressions such as R.zeros([16], "int32") would be extracted out into the parameter transformation, even though they do not depend on any parameters.
`
Does this affect the result? If a weight transformation depends on some values like `R.zeros`, such transformation will no longer be lifted if `R.zeros` is not lifted, maybe better we can check such dependency

**Comment by Lunderberg** [comment]
> Does this affect the result? If a weight transformation depends on some values like `R.zeros`, such transformation will no longer be lifted if `R.zeros` is not lifted, maybe better we can check such dependency

Good question, and this case should be handled, by allowing zero-param operators to potentially appear in both functions.  (See [this unit test](https://github.com/apache/tvm/pull/16595/files#diff-0164f466641405b90811062a983b85caf8118561d094a7c89515922d1c34044fR822) for how this looks in practice.)  While the case isn't ever explicitly handled, it instead results from the overall lifting.

1. Add every variables binding that doesn't require runtime parameters to the lifted `transform_params` function.
2. For every variable binding that is present in `transform_params` and depends on the model weights, replace with the output from `transform_params`.
3. Run dead-code elimination on both functions.

Since the zero-param operators don't depend on runtime parameters, they appear in the `transform_params`.  Since the zero-param operators don't depend on the model weights, they aren't replaced in the original function.  If they are only required in one of the two functions, then the dead-code elimination will remove them from the other.

File: src/relax/transform/lift_transform_params.cc
Comment (slyubomirsky) [inline src/relax/transform/lift_transform_params.cc:344]: Is this the logic that is meant to handle zero-param operators? I had a really hard time figuring out what the change was from #16594 and this was the only major area that came up in the diff.

**Review by slyubomirsky — COMMENTED** [review]
I realize it's a draft but I had a look anyway. Since the code changes also included those from #16594, it was a little difficult to see what had changed. I didn't see anything to take issue with, though it was a little less than obvious to see how the delta from #16594 accomplished the stated purpose, but I think I follow it.

File: src/relax/transform/lift_transform_params.cc
Comment (slyubomirsky) [inline src/relax/transform/lift_transform_params.cc:204]: Question separate from the review: Are we going with the convention of having member var names end in an underscore? I'm not a partisan on that one, but we should be consistent. @tqchen

**Comment by Lunderberg** [comment]
> I realize it's a draft but I had a look anyway. Since the code changes also included those from https://github.com/apache/tvm/pull/16594, it was a little difficult to see what had changed.

Apologies there.  Once #16594 lands, the "Files Changed" tab in this PR should update to only show the changes unique to this PR.  In the meantime, this PR branch has its changes in a separate commit ([link](https://github.com/apache/tvm/pull/16595/commits/a8ca845a883cfa7440941cba5ec35c11e2294bc8)), where they can be viewed separately from the #16594 changes.

File: include/tvm/relax/expr.h
Comment (Lunderberg) [reply include/tvm/relax/expr.h:172]: The `MarkGraphNode` function should be called for IR nodes that have reference equality, such as variables.  Using it with the `relax::CallNode` means that results from `StructuralEqual` will be dependent on how a `relax::Call` node was constructed, even if they have identical contents.

Relevant to this PR, if `R.zeros([16], "int32")` appears in both `main` and `transform_params`, the expected output generated by the TVMScript parser would have two different `relax::Call` objects, while the output of `LiftTransformParams` would use the same `relax::Call` object.  Because the `relax::Call` objects were being checked for analogous reference equality, this would cause `StructuralEqual` to erroneously report the test as failing.

I've added a unit test to specifically exercise this behavior, rather than implicitly relying on the tests for `LiftTransformParams`.

File: src/relax/transform/lift_transform_params.cc
Comment (Lunderberg) [reply src/relax/transform/lift_transform_params.cc:344]: This section along with [this line](https://github.com/apache/tvm/pull/16595/commits/a8ca845a883cfa7440941cba5ec35c11e2294bc8#diff-dbc6646d7cc42ae403e97adb88f092c13fd91bc928595b7f44d54a3f218d714bR124), yes.  This section collects information to determine if this parameter depends, directly or indirectly, on one of the model weights.  The linked line uses the collected information to determine if the `transform_params` should output a variable.

I've updated the comment here to indicate why the additional information is being collected.

File: src/relax/transform/lift_transform_params.cc
Comment (Lunderberg) [reply src/relax/transform/lift_transform_params.cc]: The TVM convention, as I understand it, is to have private members have a trailing underscore, while public members do not.  Often if I'm making quick subclasses of `ExprMutator`, I'll have a struct with public fields rather than a class with private fields.  Since the class definition already occurs within a function scope, the visibility of the entire class is already restricted, and additional visibility restrictions of private members isn't necessary.

That said, I've updated it to be a class with a private `to_suppress_`.  I figure that if a deviation from convention is big enough to raise a question, it's big enough to avoid the deviation altogether.

File: include/tvm/relax/expr.h
Comment (slyubomirsky) [reply include/tvm/relax/expr.h:172]: Good catch!

File: src/relax/transform/lift_transform_params.cc
Comment (slyubomirsky) [reply src/relax/transform/lift_transform_params.cc]: I was just curious, I don't think it's worthy of holding up a review or anything like that. Having a different rule for public/private members is reasonable.

**Comment by slyubomirsky** [comment]
Thank you for the changes. The new unit test and the new comment are both helpful.

**Comment by Lunderberg** [comment]
With #16594 landed, I've rebased this PR on top of it, and it is now ready for review.

**Review by slyubomirsky — APPROVED** [review]
Thanks for responding to previous feedback.

**Pull request merged**
